### PR TITLE
Added getCurrentRobotState method to address Issue #3, #4 and #10

### PR DESCRIPTION
**Pull request by nsnitish**
Separated the current_robot_state_ (stores the current position of the robot) from robot_state_ (used for FK and IK calls)

### CODE REVIEW

**Comment by nsnitish** [comment]
@shaun-edwards ! Can you please have a look?

File: moveit_simple/include/moveit_simple/moveit_simple.h [hunk 1]
```diff
@@ -226,6 +226,15 @@ class Robot
...
                                   double t) const;
 
+
+  /**
+  * @brief getCurrentRobotState - Returns a vector<double> of the
+  * current joint positions of the robot from robot_state_.
+  *
+  * @return std::vector<double> current_joint_position
+  */
+  std::vector<double> getCurrentRobotState(void) const;
```
Comment (shaun-edwards) [inline moveit_simple/include/moveit_simple/moveit_simple.h]: Let's call this `getJointState`.  I know MoveIt calls it `RobotState`, but that's confusing to a more traditional robot programmer.

File: moveit_simple/include/moveit_simple/moveit_simple.h [hunk 2]
```diff
@@ -315,12 +324,14 @@ class Robot
 
   void updateState(const sensor_msgs::JointStateConstPtr& msg);
 
+  void updateCurrentState(const sensor_msgs::JointStateConstPtr& msg);
 
   // Robot internal objects
   std::map<std::string, TrajectoryInfo> traj_info_map_;
 
   // MoveIt objects
   mutable moveit::core::RobotStatePtr robot_state_;
```
Comment (shaun-edwards) [inline moveit_simple/include/moveit_simple/moveit_simple.h]: I realize this is the name that I chose, but now that there is two "robot states", the names should indicate the difference.  That isn't the case right now.  Can you think of new names?

File: moveit_simple/src/moveit_simple.cpp [hunk 3]
```diff
@@ -770,6 +770,21 @@ void Robot::updateState(const sensor_msgs::JointStateConstPtr& msg)
...
+  std::lock_guard<std::recursive_mutex> guard(m_);
+  ros::spinOnce();
+  std::vector<double> current_joint_positions;
+  current_robot_state_->update();
+  current_robot_state_->copyJointGroupPositions(joint_group_->getName(), current_joint_positions);
+  return current_joint_positions;
+}
+
+
+void Robot::updateCurrentState(const sensor_msgs::JointStateConstPtr& msg)
```
Comment (shaun-edwards) [inline moveit_simple/src/moveit_simple.cpp]: Why not stick to a `vector<double>`?  I prefer to not pollute this class with ROS message types except where necessary.

File: moveit_simple/test/moveit_simple.cpp [hunk 4]
```diff
@@ -24,6 +24,23 @@ using testing::Types;
 namespace moveit_simple_test
 {
 
+
+std::string vector_to_string(const std::vector<double> &input_vector)
```
Comment (shaun-edwards) [inline moveit_simple/test/moveit_simple.cpp]: This should not be needed.  The pretty print header takes care of this.

File: moveit_simple/test/moveit_simple.cpp [hunk 5]
```diff
@@ -289,10 +306,10 @@ TEST(MoveitSimpleTest, interpolation)
   for (std::size_t i = 0; i < joint_expected.size(); ++i)
   {
     error_joint += fabs(joint_out[i] - joint_expected[i]);
-    ROS_INFO_STREAM(" joint_out_ " << i << joint_out[i]);
-    ROS_INFO_STREAM(" joint_expected_ " << i << joint_expected[i]);
   }
   EXPECT_NEAR(error_joint, 0.0, 1e-2);
+  ROS_INFO_STREAM("joint_out" << vector_to_string(joint_out));
```
Comment (shaun-edwards) [inline moveit_simple/test/moveit_simple.cpp]: Have you tried using just the stream operator, `<<`?  The prettyprint header should automatically make this work.

File: moveit_simple/test/moveit_simple.cpp [hunk 6]
```diff
@@ -419,4 +435,49 @@ TEST(MoveitSimpleTest, kinematics)
 
 }
 
+
+TEST(MoveitSimpleTest, copy_current_pose)
+{
+  moveit_simple::Robot robot(ros::NodeHandle(), "robot_description", "manipulator");
```
Comment (shaun-edwards) [inline moveit_simple/test/moveit_simple.cpp]: This test and others has a lot of the same setup.  Look into Gtest Fixtures to eliminate some of this common code.

**Review by shaun-edwards — COMMENTED** [review]
Couple of issues to address.  Also, the travis CI is failing.

File: moveit_simple/include/moveit_simple/moveit_simple.h (hunk 2 above)
Comment (nsnitish) [reply moveit_simple/include/moveit_simple/moveit_simple.h]: One says robot_state_ and other one is called current_robot_state_. Should we name them like online_robot_state_ and virtual_robot_state_?

File: moveit_simple/src/moveit_simple.cpp (hunk 3 above)
Comment (nsnitish) [reply moveit_simple/src/moveit_simple.cpp]: @shaun-edwards! You are referring to not using the ROS messages in [updateState](https://github.com/shaun-edwards/moveit_simple/pull/21/files#diff-8ae580aedd7e6266898c159ae2e4a380R767), right?

**Comment by nsnitish** [comment]
@shaun-edwards! Not sure why travis CI didn't run test this time!

File: moveit_simple/src/moveit_simple.cpp (hunk 3 above)
Comment (shaun-edwards) [reply moveit_simple/src/moveit_simple.cpp]: yep.

**Comment by shaun-edwards** [comment]
@nsnitish, please resolve conflicts

**Comment by nsnitish** [comment]
Please don't merge this PR as of now. Still need to update few things.

File: moveit_simple/src/moveit_simple.cpp (hunk 3 above)
Comment (nsnitish) [reply moveit_simple/src/moveit_simple.cpp]: Just realized, we just need only current state to be updated through the NodeHandle subscribe.

**Comment by nsnitish** [comment]
@shaun-edwards! Please have a look.

File: moveit_simple/include/moveit_simple/moveit_simple.h [hunk 7]
```diff
@@ -353,15 +341,68 @@ class Robot
...
+
+protected:
+  OnlineRobot();
+
+  void updateCurrentState(const sensor_msgs::JointStateConstPtr& msg);
+  mutable moveit::core::RobotStatePtr current_robot_state_;
+
+
+  // Visualizations
+  moveit_visual_tools::MoveItVisualToolsPtr visual_tools_;
```
Comment (shaun-edwards) [inline moveit_simple/include/moveit_simple/moveit_simple.h]: Should visual tools be in the `Robot` class?  If I remember correctly, this object provides a visual or the robot state.  Maybe we should have one for the 'real' robot and one for the 'virtual' robot.

File: moveit_simple/src/moveit_simple.cpp [hunk 8]
```diff
@@ -799,14 +803,34 @@ trajectory_msgs::JointTrajectoryPoint Robot::toJointTrajPtMsg(
...
+  std::vector<double> current_joint_positions;
+  virtual_robot_state_->update();
+  virtual_robot_state_->copyJointGroupPositions(joint_group_->getName(), current_joint_positions);
+  return current_joint_positions;
+}
+
+
 
-void Robot::updateState(const sensor_msgs::JointStateConstPtr& msg)
+std::vector<double> OnlineRobot::getJointState(void) const
```
Comment (shaun-edwards) [inline moveit_simple/src/moveit_simple.cpp:872]: Since this method overries the `Robot::getJointState` do we lose access to the virtual robot joint position?  Is this what you intended?

File: moveit_simple/test/moveit_simple.cpp [hunk 9]
```diff
@@ -25,23 +25,78 @@ using testing::Types;
...
 TEST(MoveitSimpleTest, reachability)
 {
-  moveit_simple::Robot robot(ros::NodeHandle(), "robot_description", "manipulator");
-  const Eigen::Affine3d pose = Eigen::Affine3d::Identity(); 
-  ros::Duration(2.0).sleep();  //wait for tf tree to populate
+  moveit_simple::Robot robot("robot_description", "manipulator");
 
   ROS_INFO_STREAM("Testing reachability of unknown point, should fail");
   EXPECT_FALSE(robot.isReachable("unknown_name"));
-  EXPECT_FALSE(robot.isReachable(pose, "random_link"));
```
Comment (shaun-edwards) [inline moveit_simple/test/moveit_simple.cpp]: Why were these deleted?

File: moveit_simple/test/moveit_simple.cpp [hunk 10]
```diff
@@ -25,23 +25,78 @@ using testing::Types;
 namespace moveit_simple_test
 {
 
+/**
+ * @brief UserRobotTest is a fixture for testing public methods.
+ * Objects that can be directly used inside the test
+ - robot pointer for moveit_simple::OnlineRobot
+*/
+class UserRobotTest : public ::testing::Test
```
Comment (shaun-edwards) [inline moveit_simple/test/moveit_simple.cpp:33]: +1 for test fixtures!

**Review by shaun-edwards — COMMENTED** [review]
Pretty good!...just a few comments/questions.

File: moveit_simple/src/moveit_simple.cpp (hunk 8 above)
Comment (nsnitish) [reply moveit_simple/src/moveit_simple.cpp:872]: virtual robot joint position will be there but that will be used for running the IK and FK. When we have online robot, seed conditions are taken from online robot joint position but otherwise virtual robot joint positions should be used.

File: moveit_simple/test/moveit_simple.cpp (hunk 9 above)
Comment (nsnitish) [reply moveit_simple/test/moveit_simple.cpp]: My bad. I'll re-add them.

File: moveit_simple/include/moveit_simple/moveit_simple.h (hunk 7 above)
Comment (nsnitish) [reply moveit_simple/include/moveit_simple/moveit_simple.h]: visual_tools_ uses `NodeHandle` to publish but `Robot` class doesn't have any `NodeHandle`. Isn't this required? Why do we need one for 'virtual' robot?

Comment (shaun-edwards) [reply moveit_simple/include/moveit_simple/moveit_simple.h]: It visualize the robot state, which `Robot` has.  It's useful for debugging.  Although it's not explicit, the MoveIt `RobotState` already requires a `NodeHandle`, it's just hidden.

File: moveit_simple/src/moveit_simple.cpp (hunk 8 above)
Comment (shaun-edwards) [reply moveit_simple/src/moveit_simple.cpp:872]: Is this a publc method?  The inheritance rules on this are confusing for most developers...I recommend a clear separation so people know what they are calling.

Comment (nsnitish) [reply moveit_simple/src/moveit_simple.cpp:872]: I think making the getJointState a protected method in Robot and public in OnlineRobot makes more sense. Because JointState in Robot doesn't make much sense for user.

File: moveit_simple/include/moveit_simple/moveit_simple.h (hunk 7 above)
Comment (nsnitish) [reply moveit_simple/include/moveit_simple/moveit_simple.h]: Do we also need separate NodeHandles for online and offline robot?

Comment (shaun-edwards) [reply moveit_simple/include/moveit_simple/moveit_simple.h]: Node handles are singletons...no matter how many you create, there is only ever one.  It's safe to the use the same one for both.

File: moveit_simple/src/moveit_simple.cpp (hunk 8 above)
Comment (shaun-edwards) [reply moveit_simple/src/moveit_simple.cpp:872]: +1.

**Comment by nsnitish** [comment]
@shaun-edwards! Please have a look

**Review by shaun-edwards — APPROVED** [review]
